### ISSUE TITLE
`0x6a80` error while signing a psbt on nanos via HWI

### ISSUE THREAD

**Issue by St333p** [LedgerHQ/app-bitcoin-new#201]
I'm facing a `0x6a80` error while signing a psbt via HWI, I can reproduce both using a physical device and on speculos, with app version 2.2.0-alpha for nanos in testnet. I understand the error indicates that some of the data contained in the psbt is incorrect, but I've not been able to pin down the exact issue yet. The wallet registration phase looks seems to be completing correctly.

I've also not been able to build/run the application in debug mode so to gather information from the `PRINTF`s in the application code. I built the app with `DEBUG=1`, but there's still no `dbg` on the device's screen and only apdus are printed, which are not really useful to me.

I am aware of https://github.com/bitcoin-core/HWI/issues/605 and my psbt seems to comply to all alction points described there: it has `'` in derivation paths and it contains global xpubs. 

CAn anybody help me pin down what is wrong with this transaction? Is there a psbt validator that enforces the same rules as the bitcoin app for ledger that might return more information on the error? 

### How to reproduce
- psbt that causes the signing process to fail:
```
cHNidP8BAF4CAAAAAeA2qlp6M7bO0FZpIy7GjjBjVoXculKgDxfUX1KErcUCAQAAAAAAAAAAAaI0AAAAAAAAIgAgZIe0JHQJPOBpJP/Te8WtPmjDmgpkNyhJUPi+y8lcjqAAAAAATwEENYfPAxSgaRGAAAAADsAO6p0E/PAyFjivhG17lCBfShzzyyUWWGffEJW9ybACYVP7IeFVCMbqCOV9k5nw8UxjSluAgPCY4RqC/X5g1koQ/IuVhywAAIABAACAAAAAgE8BBDWHzwMzgQi6gAAAADofW/IFBM4bInw3fP0KQAZQ4Ois914QG7DD/Ldinl9uA2vYIN86ZcfrsiOelxTQuHa8w53K7jKXCiX+05P0fmizEKR3QrMsAACAAQAAgAAAAIBPAQQ1h88DO24vCIAAAAAF2z+SxE6Y1Z84+Pijh82EAOYSPqOj3WnJmcDjsmkkiQO9UqMQcFs+B763xzuKMHxkfd8Ob2HBXqHeNre2j71pZRAfrs+DLAAAgAEAAIAAAACATwEENYfPAz0qT/KAAAAABQwr0X1+jhp8YR23B9gu2zHoGA+M+p+zvKB4W0CzFgoC9BHxK6mS0/QygHkZ4C4+P6n4WtWOa18Yapnv11nEq1cQymDkRiwAAIABAACAAAAAgE8BBDWHzwNb6TMLgAAAAJIJavg9NYqlh5lfx6Iz2koV+oVlo8RFjB8k4M0Y704TAm4qHEoKXflmgQ5yLdrHZTgLT/aQTlFBEjJ6Gx+YZo7kEKN/FzssAACAAQAAgAAAAIBPAQQ1h88DaW19jYAAAAA8WhyX6XBUUG9r2tr7Cy5cftKU0ddT0GZHUu8pUSwBPgPeGHFnlhmoSv3Duxj+CBHncf+mcPQxOwCe/EC0mYPoIhAN8rafLAAAgAEAAIAAAACATwEENYfPA3nVHSmAAAAAEfE6V5mNTtchOnbi3UKEY/zbWDAyKpdLxvR0UCelYmADtLYbOzSdV5PEbFLCwWyUFD1qWw7T+y5S4Ol2Vc6Ve2UQ2SkKECwAAIABAACAAAAAgE8BBDWHzwOli8uEgAAAAIjFOppqxl3abp0pQI4w29Nqtf7WZESeHbkvdQWWNFQBAgJgAL+EJa4WSsM+Zd8l6vR0wZ0Cd0XIQw4A3boWKchgECXzuJssAACAAQAAgAAAAIBPAQQ1h88DpZJEd4AAAAB9PlRDehI4H+Y2w5Iq3oW2QY1hRxwBzYCU85O0Qxs30wMJ6OYlj/7lYU6AGuSVUeWQLFWgk3vBS0f3QlF1RP4flRCTuO6HLAAAgAEAAIAAAACAAAEAfQIAAAAB+RIUzdB1Hi/6Zwn5XLX1L0aMFPsWXKvZBYOJQLPM6J8BAAAAAP3///8Ctx8bAAAAAAAWABRuUAEzivL9fI5eoD6FsQEG5D+Drcg1AAAAAAAAIgAgZIe0JHQJPOBpJP/Te8WtPmjDmgpkNyhJUPi+y8lcjqCVtyYAAQEryDUAAAAAAAAiACBkh7QkdAk84Gkk/9N7xa0+aMOaCmQ3KElQ+L7LyVyOoAEDBAEAAAABBf2oAVIhAhQsyd4TgjV0hMmZtdqTTSEGX7tx2pGYLM13tnjveKZKIQJfFzxdXAFlY84F/bllPBsjAAbOZiFj0uISGMMQyWkhVCECkFxOiO+GhoZfXVCCPnD1UjzYsiNbHkL7FMUuhmlcS71TrmRSIQLFsg5KEJ+wsSO3PlxcDM77yuIZRSz2IhbvvwzA0g627iEDfwgDIDXeu2GMSAa9uAy9e9zFjohcAca6e6AYkbCeP5UhA/lSTepxylP/yJaJYc17dQGQ8jYeDrOqeKME0p8OrP2jU68CuACyZ1MhAlAtbyC5FXO8IiF2j/gjO47TIlRAcbDKqDQT2hSMKMPCIQKtixjp2H9xVMwH3M+NzCCsbj6ulZo7lJdVyAU5Oqp4nyECv10fiWan+I1tRqQ1XjPsKIBIEdc6ciWM1+22NAu+KIMhAzd4ybAaOLOd32/xlDxn2z4o7mz1ekOMj8xCcQzrs4DUIQNYVRC7x8Ogp9o77nwfHUrZ+BVcaELPal1s7fcswBWzhiEDpOyajeNHLsuhLWrE2bsAiZZkom8mOgtA7h/j7OUZBQpWrmgiBgIULMneE4I1dITJmbXak00hBl+7cdqRmCzNd7Z473imShgN8rafLAAAgAEAAIAAAACAAQAAAAAAAAAiBgJQLW8guRVzvCIhdo/4IzuO0yJUQHGwyqg0E9oUjCjDwhijfxc7LAAAgAEAAIAAAACAAQAAAAAAAAAiBgJfFzxdXAFlY84F/bllPBsjAAbOZiFj0uISGMMQyWkhVBj8i5WHLAAAgAEAAIAAAACAAQAAAAAAAAAiBgKQXE6I74aGhl9dUII+cPVSPNiyI1seQvsUxS6GaVxLvRgfrs+DLAAAgAEAAIAAAACAAQAAAAAAAAAiBgKtixjp2H9xVMwH3M+NzCCsbj6ulZo7lJdVyAU5Oqp4nxjKYORGLAAAgAEAAIAAAACAAQAAAAAAAAAiBgK/XR+JZqf4jW1GpDVeM+wogEgR1zpyJYzX7bY0C74ogxgl87ibLAAAgAEAAIAAAACAAQAAAAAAAAAiBgM3eMmwGjiznd9v8ZQ8Z9s+KO5s9XpDjI/MQnEM67OA1BiTuO6HLAAAgAEAAIAAAACAAQAAAAAAAAAiBgNYVRC7x8Ogp9o77nwfHUrZ+BVcaELPal1s7fcswBWzhhikd0KzLAAAgAEAAIAAAACAAQAAAAAAAAAiBgOk7JqN40cuy6EtasTZuwCJlmSibyY6C0DuH+Ps5RkFChjZKQoQLAAAgAEAAIAAAACAAQAAAAAAAAAAAA==
```
- `speculos` cmd for `1faecf83`: 
```
speculos build/nanos/bin/app.elf -m nanos -s "print verb ill rubber junk people virtual lyrics pole grain style reduce ocean tenant step divide neutral voice resist song east unable result rabbit"
``` 
- Signing as `hwi -f 1faecf83 --chain test signtx <base64_psbt_above>`
- hwi version 2.3.1
- ledger-bitcoin app version 2.2.0-alpha

**Comment by bigspider** [comment]
I don't see anything obviously wrong with PSBT, except that it's a miniscript policy, and miniscript signing is not implemented in HWI. I made a [PoC](https://github.com/bitcoin-core/HWI/pull/647) (which is probably now somewhat outdated, but it should still work from that branch), but there is no recent work on that issue, and I would probably do things differently today.

[async-hwi](https://github.com/wizardsardine/async-hwi) is the first library trying to integrate this in a vendor-agnostic way, but it's Rust.

You should be able to sign such a PSBT using one of the client libraries of the Ledger bitcoin app.

**Comment by St333p** [comment]
I looked a bit into how HWI handles the signing process and I'm not sure what I should do differently.

I'm assuming the wallet policy is registered correctly, since the ledger app gets to the "spend from known wallet" phase. At that point, the line in HWI that fails is just:
```
            input_sigs = self.client.sign_psbt(psbt2, wallet, wallet_hmac)
```
If I were to implement the signing process using ledger's python client library I'd probably do it the same way. So to me it seems like at least one of the following is true:
- `ledger_bitcoin.NewClient.sign_psbt` is the wrong method for the use case or it's called with wrong parameters
- There is something wrong in the wallet policy, so that the registration does not fail but signing fails using that policy.

In particular, even after skimming through your PoC PR, it's still not clear to me whether miniscript-specific logic should be applied at the wallet registration phase or at the psbt signing phase.

Note: I'll add more information to the issue description in case anyone wants to reproduce

**Comment by bigspider** [comment]
HWI has no concept of "wallet policies", so it uses the information in the psbt to reconstruct it; this was only implemented for multisig.
In principle, it could be generalized to miniscript on SegWit - but it's impossible to generalize it to generic taproot scripts, since the psbt fields that are defined today only contain info about the leaf that is being spent.

In the POC PR, I add APIs to specify explicitly the wallet policy as an argument (`sign(psbt, wallet_policy, wallet_policy_hmac)` instead of just `sign(psbt)`; an alternative would be to add the wallet policy / descriptor somewhere in the PSBT, but there's currently no standard defined for that).

So in short: there is currently no way to sign scripts more complex than multisig with HWI; you can do it using the ledger-provided libraries.

**Comment by St333p** [comment]
I guess what confused me is that HWI is able to successfully register a wallet policy based on the psbt. If complex scripts are not supported, I'd expect it to fail there and not in the signing phase.

Thanks for your time
